Error in Array compilation

### Describe the bug

All of my code had been working fine, I had created an intermediary type:

```v

struct Tensor {
    data []f64
    len int
}
```

And decided to replace it directly with []f64s, the compiler now returns `error: cannot convert 'void *' to 'struct array'`

### Expected Behavior

For the code to compile

### Current Behavior

```
src/main.v:3:8: warning: module 'ai (src.ai)' is imported but never used
    1 | module main
    2 | 
    3 | import ai
      |        ~~
    4 | 
    5 | fn main() {
==================
/tmp/v_1000/AI.7046668324025237048.tmp.c:25543: error: cannot convert 'void *' to 'struct array'
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang
```

### Reproduction Steps

You can get the code on my repository: https://github.com/neosapien3247/vai

The previous verison compiled without issue: ede4f523382600b6ea1a1005fff6210962683388

However it now no longer compiles

### Possible Solution

_No response_

### Additional Information/Context

_No response_

### V version

```bash v --version V 0.3.3 f08b882 ```

### Environment details (OS name and version, etc.)

OS: elementaryOS 7 on an x86_64 architecture

Built on Ubuntu 22.04.2 LTS
Linux 5.19.0-35-generic

Okay, the error comes from a wrong cast:

```v
// test.vsh

import arrays

x := [1.0, 2.0, 3.0, 4.0]

println(arrays.sum[f64](x.data)!)
```

This, instead of returning an error about mismatched types between an []f64 and a voidptr (since the `data` field is a pointer to the array) It returns a `builder error` with a C compilation error:

```bash
==================
/tmp/v_1000/test.16293055329910315485.tmp.c:17714: error: cannot convert 'void *' to 'struct array'
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang
```

Ive narrowed down the problem, the issue doesn't seem to be with the generics, rather it is simply a failure for checking the type of a `voidptr`:

```v
fn main() {
	c := voidptr(10)

	// println(sum[i32](c)!)
	println(test(c))
}

fn test(a []i32) i32 {
	return a[0]
}
```

returns:

```bash
/tmp/v_1000/hello.1476530331192383116.tmp.c:12475: error: cannot convert 'void *' to 'struct array'
builder error: 
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang
```

what do you think could be causing this @felipensp ?

Looking at the definition of the array getter method and its corresponding C translation, I wonder if it might have to do with an implicit cast on the return line:

```v
//vlib/builtin/array.v:443
// Private function. Used to implement x = a[i] or { ... }
fn (a array) get_with_check(i int) voidptr {
	if i < 0 || i >= a.len {
		return 0
	}
	unsafe {
		return &u8(a.data) + u64(i) * u64(a.element_size) // <-- here
	}
}
```
```c
voidptr array_get(array a, int i) {
	#if !defined(CUSTOM_DEFINE_no_bounds_checking)
	{
		if (i < 0 || i >= a.len) {
			panic_debug(435, tos3("/home/neosapien/.local/bin/vlib/builtin/array.v"), tos3("builtin"), tos3("get"),   str_intp(3, _MOV((StrIntpData[]){{_SLIT("array.get: index out of range (i == "), /*100 &int*/0xfe07, {.d_i32 = i}}, {_SLIT(", a.len == "), /*100 &int*/0xfe07, {.d_i32 = a.len}}, {_SLIT(")"), 0, { .d_c = 0 }}})));
			VUNREACHABLE();
		}
	}
	#endif
	{ // Unsafe block
		return ((u8*)(a.data)) + ((u64)(i)) * ((u64)(a.element_size));
	}
	return 0;
}
```

--- a/vlib/v/checker/check_types.v
+++ b/vlib/v/checker/check_types.v
@@ -306,7 +306,7 @@ fn (mut c Checker) check_expected_call_arg(got ast.Type, expected_ ast.Type, lan
 	if c.check_types(got, expected) {
 		if language == .v && idx_got == ast.voidptr_type_idx {
 			if expected.is_int_valptr() || expected.is_int() || idx_expected == ast.voidptr_type_idx
-				|| idx_expected == ast.charptr_type_idx {
+				|| idx_expected == ast.charptr_type_idx || idx_expected == ast.byteptr_type_idx {
 				return
 			}
 			exp_sym := c.table.final_sym(expected)

--- a/vlib/v/checker/check_types.v
+++ b/vlib/v/checker/check_types.v
@@ -162,7 +162,7 @@ fn (mut c Checker) check_types(got ast.Type, expected ast.Type) bool {
 			return true
 		}
 		exp_sym := c.table.final_sym(expected)
-		if exp_sym.kind !in [.function, .placeholder] {
+		if exp_sym.kind !in [.voidptr, .function, .placeholder] {
 			return false
 		}
 	}

--- a/vlib/v/checker/check_types.v
+++ b/vlib/v/checker/check_types.v
@@ -158,7 +158,7 @@ fn (mut c Checker) check_types(got ast.Type, expected ast.Type) bool {
 		}
 	}
 	if got_idx == ast.voidptr_type_idx {
-		if expected.is_int_valptr() {
+		if expected.is_int_valptr() || expected.is_int() {
 			return true
 		}
 		exp_sym := c.table.final_sym(expected)

--- a/vlib/v/checker/check_types.v
+++ b/vlib/v/checker/check_types.v
@@ -305,12 +305,13 @@ fn (mut c Checker) check_expected_call_arg(got ast.Type, expected_ ast.Type, lan
 
 	if c.check_types(got, expected) {
 		if language == .v && idx_got == ast.voidptr_type_idx {
-			if expected.is_int_valptr() || expected.is_int() {
+			if expected.is_int_valptr() || expected.is_int() || idx_expected == ast.voidptr_type_idx
+				|| idx_expected == ast.charptr_type_idx {
 				return
 			}
 			exp_sym := c.table.final_sym(expected)
 			if exp_sym.language == .v
-				&& exp_sym.kind !in [.voidptr, .function, .placeholder, .array_fixed, .struct_] {
+				&& exp_sym.kind !in [.voidptr, .charptr, .byteptr, .function, .placeholder, .array_fixed, .struct_] {
 				got_typ_str, expected_typ_str := c.get_string_names_of(got, expected)
 				return error('cannot use `${got_typ_str}` as `${expected_typ_str}`')
 			}

--- a/vlib/v/checker/check_types.v
+++ b/vlib/v/checker/check_types.v
@@ -310,7 +310,7 @@ fn (mut c Checker) check_expected_call_arg(got ast.Type, expected_ ast.Type, lan
 			}
 			exp_sym := c.table.final_sym(expected)
 			if exp_sym.language == .v
-				&& exp_sym.kind !in [.voidptr, .function, .placeholder, .array_fixed] {
+				&& exp_sym.kind !in [.voidptr, .function, .placeholder, .array_fixed, .struct_] {
 				got_typ_str, expected_typ_str := c.get_string_names_of(got, expected)
 				return error('cannot use `${got_typ_str}` as `${expected_typ_str}`')
 			}

--- a/vlib/v/checker/check_types.v
+++ b/vlib/v/checker/check_types.v
@@ -162,7 +162,7 @@ fn (mut c Checker) check_types(got ast.Type, expected ast.Type) bool {
 			return true
 		}
 		exp_sym := c.table.final_sym(expected)
-		if exp_sym.kind !in [.voidptr, .function, .placeholder] {
+		if exp_sym.kind !in [.voidptr, .function, .placeholder, .array_fixed] {
 			return false
 		}
 	}

--- a/vlib/v/checker/check_types.v
+++ b/vlib/v/checker/check_types.v
@@ -305,8 +305,7 @@ fn (mut c Checker) check_expected_call_arg(got ast.Type, expected_ ast.Type, lan
 
 	if c.check_types(got, expected) {
 		if language == .v && idx_got == ast.voidptr_type_idx {
-			if expected.is_int_valptr() || expected.is_int() || idx_expected == ast.voidptr_type_idx
-				|| idx_expected == ast.charptr_type_idx || idx_expected == ast.byteptr_type_idx {
+			if expected.is_int_valptr() || expected.is_int() || expected.is_ptr() {
 				return
 			}
 			exp_sym := c.table.final_sym(expected)

--- a/vlib/v/checker/check_types.v
+++ b/vlib/v/checker/check_types.v
@@ -157,15 +157,6 @@ fn (mut c Checker) check_types(got ast.Type, expected ast.Type) bool {
 			return true
 		}
 	}
-	if got_idx == ast.voidptr_type_idx {
-		if expected.is_int_valptr() || expected.is_int() {
-			return true
-		}
-		exp_sym := c.table.final_sym(expected)
-		if exp_sym.kind !in [.voidptr, .function, .placeholder, .array_fixed] {
-			return false
-		}
-	}
 	if expected == ast.charptr_type && got == ast.char_type.ref() {
 		return true
 	}
@@ -313,6 +304,17 @@ fn (mut c Checker) check_expected_call_arg(got ast.Type, expected_ ast.Type, lan
 	}
 
 	if c.check_types(got, expected) {
+		if language == .v && idx_got == ast.voidptr_type_idx {
+			if expected.is_int_valptr() || expected.is_int() {
+				return
+			}
+			exp_sym := c.table.final_sym(expected)
+			if exp_sym.language == .v
+				&& exp_sym.kind !in [.voidptr, .function, .placeholder, .array_fixed] {
+				got_typ_str, expected_typ_str := c.get_string_names_of(got, expected)
+				return error('cannot use `${got_typ_str}` as `${expected_typ_str}`')
+			}
+		}
 		if language != .v || expected.is_ptr() == got.is_ptr() || arg.is_mut
 			|| arg.expr.is_auto_deref_var() || got.has_flag(.shared_f)
 			|| c.table.sym(expected_).kind !in [.array, .map] {

--- a/vlib/v/checker/check_types.v
+++ b/vlib/v/checker/check_types.v
@@ -157,6 +157,15 @@ fn (mut c Checker) check_types(got ast.Type, expected ast.Type) bool {
 			return true
 		}
 	}
+	if got_idx == ast.voidptr_type_idx {
+		if expected.is_int_valptr() {
+			return true
+		}
+		exp_sym := c.table.final_sym(expected)
+		if exp_sym.kind !in [.function, .placeholder] {
+			return false
+		}
+	}
 	if expected == ast.charptr_type && got == ast.char_type.ref() {
 		return true
 	}

--- a/vlib/v/checker/tests/voidptr_arg_err.out
+++ b/vlib/v/checker/tests/voidptr_arg_err.out
@@ -1,0 +1,7 @@
+vlib/v/checker/tests/voidptr_arg_err.vv:3:15: error: cannot use `voidptr` as `[]i32` in argument 1 to `test`
+    1 | fn test_main() {
+    2 |     c := voidptr(10)
+    3 |     println(test(c))
+      |                  ^
+    4 | }
+    5 |

--- a/vlib/v/checker/tests/voidptr_arg_err.vv
+++ b/vlib/v/checker/tests/voidptr_arg_err.vv
@@ -1,0 +1,8 @@
+fn test_main() {
+	c := voidptr(10)
+	println(test(c))
+}
+
+fn test(a []i32) i32 {
+	return a[0]
+}

